api/src/test/java/org/pmiops/workbench/api/DataSetControllerTest.java

```diff
@@ -0,0 +1,419 @@
+    ValueSet valueSet = new ValueSet().domain(Domain.CONDITION).values(new DomainValuesResponse().items(domainValues));
+    dataSet.addValuesItem(valueSet);
+
+
+    List<String> selectStrings = new ArrayList<>();
+    selectStrings.add("PERSON_ID");
+    List<String> joinStrings = new ArrayList<>();
+    joinStrings.add("FROM `all-of-us-ehr-dev.synthetic_cdr20180606.condition_occurrence` c_occurrence");
+
+    dataSetController = spy(dataSetController);
+    doReturn(new ValuesLinkingPair(selectStrings, joinStrings)).when(dataSetController).getValueSelectsAndJoins(valueSet, Domain.CONDITION);
+
+    DataSetQueryResponse response = dataSetController.getQueryFromDataSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, dataSet).getBody();
+    assertThat(response.getQueryList().size()).isEqualTo(1);
+    verify(dataSetController, times(1)).getValueSelectsAndJoins(valueSet, Domain.CONDITION);
+    assertThat(response.getQueryList().get(0).getQuery()).matches("SELECT(.*)WHERE CONDITION_CONCEPT_ID IN \\((.*)\\)(.*)PERSON_ID IN \\(SELECT(.*)");
+  }
+
+  @Test
+  public void testGetQueryTwoDomains() {
+    DataSet dataSet = buildEmptyDataSet();
+    dataSet = dataSet.addCohortIdsItem(COHORT_ONE_ID);
+    dataSet = dataSet.addConceptSetIdsItem(CONCEPT_SET_ONE_ID);
+    dataSet = dataSet.addConceptSetIdsItem(CONCEPT_SET_TWO_ID);
+    List<DomainValue> domainValues = new ArrayList<>();
+    domainValues.add(new DomainValue().value("PERSON_ID"));
+    ValueSet valueSet = new ValueSet().domain(Domain.CONDITION).values(new DomainValuesResponse().items(domainValues));
+    dataSet.addValuesItem(valueSet);
+    ValueSet valueSetTwo = new ValueSet().domain(Domain.DRUG).values(new DomainValuesResponse().items(domainValues));
+    dataSet.addValuesItem(valueSetTwo);
+
+
+    List<String> selectConditionStrings = new ArrayList<>();
+    selectConditionStrings.add("PERSON_ID");
+    List<String> joinConditionStrings = new ArrayList<>();
+    joinConditionStrings.add("FROM `all-of-us-ehr-dev.synthetic_cdr20180606.condition_occurrence` c_occurrence");
+
+    List<String> selectDrugStrings = new ArrayList<>();
+    selectDrugStrings.add("PERSON_ID");
+    List<String> joinDrugStrings = new ArrayList<>();
+    joinDrugStrings.add("FROM `all-of-us-ehr-dev.synthetic_cdr20180606.drug_exposure` d_exposure");
+
+    dataSetController = spy(dataSetController);
+    doReturn(new ValuesLinkingPair(selectConditionStrings, joinConditionStrings)).when(dataSetController).getValueSelectsAndJoins(valueSet, Domain.CONDITION);
+    doReturn(new ValuesLinkingPair(selectDrugStrings, joinDrugStrings)).when(dataSetController).getValueSelectsAndJoins(valueSetTwo, Domain.DRUG);
+
+    DataSetQueryResponse response = dataSetController.getQueryFromDataSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, dataSet).getBody();
+    assertThat(response.getQueryList()).isNotEmpty();
+    verify(dataSetController, times(2)).getValueSelectsAndJoins(any(), any());
```

Comment:
Side-node / optional: I'm generally not convinced on the value of spies. The core assertions should be based on the API of the entity under test, which IMO is the below assertion. (Unless you think for some reason we're at high risk of regressions, or that checking the # of calls to this method is evaluating something entirely separate from the next line's assertion.)

Reply:
Because we're dropping queries in some places, I want to make sure we are actually fetching all the data we need, especially on joins. I agree with you, in general I don't think testing against spies are very useful (although we do need it here so that we can mock out the returns on a method in a not-mocked context)